File: multiples-of-3-and-5/elixir/multiple.ex
defmodule Multiple do
  import Enum, only: [reduce: 3]

  def multiple_of(numbers, quantity) do
    reduce(quantity..1, [],
      fn (target, accumulator) ->
        if is_multiple(target, numbers), do: [target | accumulator], else: accumulator
      end)
  end

  defp is_multiple(target, [number | numbers]) do
    is_multiple(target, number) || is_multiple(target, numbers)
  end

  defp is_multiple(_target, []), do: false

  defp is_multiple(target, number) do
    rem(target, number) == 0
  end
end

File: multiples-of-3-and-5/clojure/src/multiple.clj
(ns multiple)

(defn is-multiple [target number]
  (= (mod target number) 0))

(defn is-multiple-of-any [target numbers]
  (reduce (fn [accumulator, number] (or accumulator (is-multiple target number)))
    false numbers))

(defn filter-multiples-of [numbers]
  (fn [accumulator target]
    (if (is-multiple-of-any target numbers)
      (conj accumulator target)
      accumulator)))

(defn multiple-of [numbers quantity]
  (reduce (filter-multiples-of numbers) [] (range 1 (+ quantity 1))))
